ENH: exact p-value in stats.kendalltau() for sample sizes > 171

Konrad0 opened the pull request:
Extended the exact method for calculating the p-value in stats.kendalltau()
(originally added in #8614) to sample sizes greater than 171.
Some roundoff error is unavoidable. Tests are also added.
(All also for the masked version.)


#### Reference issue
Originally, the exact method for calculating the p-value in stats.kendalltau() was implemented in #8614. But due to overflow it only works for sample sizes up to 171.

#### What does this implement/fix?
The extension to the exact p-value calculation in stats.kendalltau() enables sample sizes greater than 171. Some roundoff is unavoidable, but as seen from the tests, the relative differences to the correct result are around 1e-14 (@chrisb83, @sturlamolden please feel free to to comment since we discussed this earlier). The correct results had to be generated using SymPy because R suffers from the same shortcomings we faced up to now. A warning is raised also because the calculation can get very slow, as the calculation has the same time complexity (O(n**3) in the worst case) as the original exact calculation.

pvanmulbregt commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1186,17 +1187,40 @@ def test_weightedtau():
+    # y[-3], y[-4] = y[-4], y[-3]
+    # _, pval = stats.kendalltau(x, y, method='exact')
+    # assert_equal(pval, 0.0)
+
```
Not needed?

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -622,7 +622,7 @@ def kendalltau(x, y, use_ties=True, use_missing=False, method='auto'):
         elif 2*c == (n*(n-1))//2:
             prob = 1.0
-        else:
+        elif n < 171:
```
If you change the cutover to something much smaller, e.g. `n=1`, does the new code return the same values as the original `n<171` code?

pvanmulbregt commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -355,6 +355,30 @@ def test_kendalltau_seasonal(self):
+                    800: (-0.018948685857321654, 0.42265448483120932055),
+                    801: (-0.014675405742821473, 0.53437553412194416236)}
+    # 1600: (0.0033270794246404007, 0.84200727400323538419),
+    # 1601: (0.01576514678326046, 0.34465255088058593946)}
```
Why are the entries for n=1600, 1601 commented out?

pvanmulbregt commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1186,17 +1187,40 @@ def test_weightedtau():
+                    400: (0.02343358395989975, 0.48444283672113314099),
+                    401: (-0.014563591022443893, 0.66363159823474837662),
+                    800: (-0.018948685857321654, 0.42265448483120932055),
+                    801: (-0.014675405742821473, 0.53437553412194416236)}
```
Are these the same values as in `test_mstats_basic.py`? If so, can you combine in any way to avoid the duplication?

pvanmulbregt replied:
One way to combine would be to factor into 3 tests: test the low-level `_kendall_p_exact`, and then tests of `stats.kendalltau` and `mstats.kendalltau` (to test the masking), which no longer would need to be as extensive.

Konrad0 replied:
Sounds good, spilt in the latest commit, so the duplication is now avoided. The input to the new test looks a bit different since it calls the p-value calculation directly, but corresponds to the cases used before.

pvanmulbregt commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -355,6 +355,30 @@ def test_kendalltau_seasonal(self):
 
+
+def test_kendall_tau_large():
+    # Test for the exact method with large samples (n > 171)
```
Should this be `n>=171` (opposite of `n<171`)?

pvanmulbregt commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -355,6 +355,30 @@ def test_kendalltau_seasonal(self):
+    # expected values generated using SymPy
+    expectations = {100: (-0.033131313131313136, 0.62822615287956040664),
+                    101: (0.03524752475247526, 0.60439525773513602669),
+                    200: (-0.015376884422110552, 0.74753983745929675209),
```
I might have expected some tests at the cutover, so n=170 and 171.

pvanmulbregt replied:
The algorithm changes at n=171.  That makes it good location to catch regressions.
Is there any test that can be done at those 2 values that would show that they exhibit some correct monotonicity behaviour? E.g. Are there pairs (170, c1) and (171, c2) where you expect the `_kendall_p_exact(...)` values to be "really close" if working correctly?

Konrad0 replied:
To get the same p-value, we would have to find two samples s1 and s2 (of size 170 and 171, respectively), such that `171*sum(new)_s1 = sum(new)_s2`. I'm not sure how to generate such samples (or the corresponding values of `c`), but I try to think of a method.

pvanmulbregt commented:
@Konrad0, It may be possible to combine the `n<171` and `n>=171` cases.  
IIUC, the concern is that the elements of `new` may overflow, and/or that `np.math.factorial(n)` is too large to convert to a float.  
One approach that handles both would be to compute `new` each iteration and then scale by a power of 2 to keep `new` within bounds.  I.e. Use `np.frexp()` on the "largest" value in `new` and then scale: `new = np.ldexp(new, -expt)`, keeping track of the total exponent. Only at the very end, check if `n<171` and then use `prob=np.ldexp(2.0*sum(new)/np.math.factorial(n), total_expt)`; or some other method (such as combining the scaling exponent with `scipy.special.gammaln(n+1)/np.log(2)`).

Konrad0 commented:
@pvanmulbregt thank you very much for your detailed review!
As mentioned above, my main motivation for the enhancement was to get rid of the wrong pvalue=0.0 for size >= 171. Actually I was hoping to do the computation in a more efficient way by using FFTs since the algorithm can be written as 
```
new = np.array([1.0])
for j in range(2,size+1):
    w = np.full(j, 1.0/j)
    new = convolve(w, new, mode='full')[:c+1]
    print(new)
pvalue = 2.0*sum(new)
```
It is possible to replace the convolution by a multiplication of the Fourier transform of each vector and perform the inverse FT at the end. Unfortunately I haven't found an efficient way of doing the calculation in Fourier space, so the O(n**3) time complexity is still unimproved, maybe you have an idea?

Anyway, I like your idea of scaling by powers of 2, I'll implement that next and update you on the results.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -526,6 +526,47 @@ def _spearmanr_2cols(x):
+        new[1] = 1.0
+        for j in range(3,n+1):
+            cs = np.cumsum(new)
+            new = cs.copy()
```
Do you need to make a copy?  I ask for two reasons.
1. The copy is only needed if `j<=c`, hence not always used.
2. Can the `new[j:] -= cs[:c+1-j]` be rewritten as `new[j:] -= new[:c+1-j]`?  This may be relying on numpy doing operations in a certain order, so may not be safe.  But `new[j:] -= new[:c+1-j].copy()` should always be safe?

Konrad0 replied:
Thanks for the suggestion, on my machine it seems to work without copy(), so I implemented it. If you have doubts that it is safe, let me know.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -548,7 +589,9 @@ def kendalltau(x, y, use_ties=True, use_missing=False, method='auto'):
         'asymptotic' uses a normal approximation valid for large samples.
         'exact' computes the exact p-value, but can only be used if no ties
-        are present. 'auto' is the default and selects the appropriate
+        are present. For sample size > 171, the exact computation may incur
```
> sample size > 171

I believe that this is " >= 171".  
And single-quotes around exact:  "the 'exact' computation may incur"

Konrad0 replied:
Changed as discussed below.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -548,7 +589,9 @@ def kendalltau(x, y, use_ties=True, use_missing=False, method='auto'):
         'exact' computes the exact p-value, but can only be used if no ties
-        are present. 'auto' is the default and selects the appropriate
+        are present. For sample size > 171, the exact computation may incur
+        roundoff error and can take very long.
```
Roundoff error may occur for `n < 171` also.

Perhaps just state as:

"As n increases, the 'exact' computation time may grow and the result may lose some precision."

Konrad0 replied:
Alright, changed (also in stats.py).

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -526,6 +526,47 @@ def _spearmanr_2cols(x):
+            new = cs.copy()
+            if j <= c:
+                new[j:] -= cs[:c+1-j]
+        prob = 2.0*sum(new)/np.math.factorial(n)
```
Do you want `sum(new)` or `np.sum(new)`?

Konrad0 replied:
Both work, so switched to `np.sum(new)`.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -526,6 +526,47 @@ def _spearmanr_2cols(x):
+        prob = 1.0
+    elif n < 171:
+        new = np.zeros(c+1)
+        new[0] = 1.0
```
`new[0:2] = 1.0`?  One line shorter and makes clear that the 1.0 is the same value twice, not 2 logically distinct values that just happen to be the same.

Konrad0 replied:
OK, changed.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -634,6 +634,24 @@ def kendalltau(x, y, use_ties=True, use_missing=False, method='auto'):
+                          "of the Kendall p-value may incur significant "
+                          "roundoff error and can take very long. Please "
+                          "consider switching to asymptotic mode.",
+                          RuntimeWarning)
```
This fixes an obviously incorrect value so agree it is worthwhile.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -634,6 +634,24 @@ def kendalltau(x, y, use_ties=True, use_missing=False, method='auto'):
+        else:
+            warnings.warn("Due to the large sample size, the exact computation "
+                          "of the Kendall p-value may incur significant "
+                          "roundoff error and can take very long. Please "
```
Was the distribution notably wider, or just the outliers?

Konrad0 replied:
The new rounding errors are generally larger and grow with the sample size as expected.

Konrad0 commented:
Hi @pvanmulbregt ,
thanks for your comments and sorry for the extended silence. I think most of your suggestions are implemented, but I just saw that several azure pipelines failed (`##[error]TF24668: The following team project collection is stopped: scipy-org. Start the collection and then try again. Administrator Reason: abuse`). Not sure what to do about that, seems like a more general issue to me.

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -526,6 +526,43 @@ def _spearmanr_2cols(x):
+            if j <= c:
+                new[j:] -= new[:c+1-j]
+        prob = np.sum(new)
+        prob = max(0.0, min(1.0, prob))
```
1. `np.clip(prob, 0, 1)` makes the intention clear.
2. This could be deindented one level and safely applied to all the branches.

Konrad0 replied:
No problem, both done.

pvanmulbregt commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -20,7 +20,7 @@
     assert_allclose, assert_array_equal)
 from numpy.testing import suppress_warnings
-
+from .. import mstats_basic
```
Rather than assuming the relative location, just import directly.
```
from scipy.stats import mstats_basic
```

Konrad0 replied:
Sure, changed.

pvanmulbregt commented:
Just a couple of small comments after those latest changes.
And the azure pipeline issue should be fixed. New commits should trigger a full set of CI testing.

Konrad0 commented:
Thanks for your comments @pvanmulbregt , they are implemented in the new commit. Hopefully the pipelines are fixed now as well.

Konrad0 commented:
Hi @pvanmulbregt and all,
looking over the code I noticed that I had removed the size checks for the special cases `c==0` and `c==1`, which threw errors for large sizes. Since it's not merged yet, I decided to fix it here. It is now reverted to its previous, correct state. Also, I added corresponding tests.
@pvanmulbregt could you take another look? Thanks!

mdhaber commented:
Hey @Konrad0 in gh-9361 both @vkk800 and I got the impression that the p-values should be the same for Tau-B and Tau-C. DO you know if that's correct, or is that your understanding of the [references](https://github.com/scipy/scipy/pull/9361#issuecomment-660195092)?

Konrad0 commented:
Hi @mdhaber ,
sorry for the delayed response. Thanks for noticing this PR, but please keep in mind that the exact method is not applicable to samples with ties. That said, the references to the SAS manual look like they're using the central limit theorem, so should be accurate only for large samples, right? I saw there was a lot of activity in the other thread, so is this issue already resolved? Thanks for all your work on this and other statistical methods!

Konrad0 commented:
@pvanmulbregt are we ready to merge here?

pvanmulbregt commented:
It needs a rebase due to merge conflicts in `stats/stats.py`

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1261,17 +1262,6 @@ def test_segfault_issue_9710():
-    y[-3], y[-4] = y[-4], y[-3]
-    _, pval = stats.kendalltau(x, y, method='exact')
-    assert_equal(pval, 0.0)
-
```
Note that this test below was added in the merge. It passes, so I kept it. Should it be removed?

mdhaber commented:
@pvanmulbregt @Konrad0 I merged (rather than rebasing and force-pushing so you can see that nothing of substance changed) upstream master. 

It sounds like this was ready from your perspective, @pvanmulbregt ?
I've had trouble getting my hands on Rank Correlation Methods. Would it be possible for you to send me the relevant parts so I can see where these changes come from?
Or if you are comfortable with it, please feel free to merge @pvanmulbregt .

mdhaber commented:
> Thanks for noticing this PR, but please keep in mind that the exact method is not applicable to samples with ties. 

Thanks. I understand, I was just checking if that was your understanding for the asymptotic methods. We decided yes, and we were pretty confident. 

> That said, the references to the SAS manual look like they're using the central limit theorem, so should be accurate only for large samples, right? 

Well, yes, but we provide them for the user to decide.

> I saw there was a lot of activity in the other thread, so is this issue already resolved? 

There was, but we were just adding tau-c.  This issue was definitely not resolved!

> Thanks for all your work on this and other statistical methods!

And thank you! This PR looks great. I haven't reviewed the code, but I have tried it out, and it's great that it extends the exact results beyond 171!

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -549,6 +549,42 @@ def _spearmanr_2cols(x):
+def _kendall_p_exact(n, c):
+    # Exact p-value, see Maurice G. Kendall, "Rank Correlation Methods" (4th Edition), Charles Griffin & Co., 1970.
+    if n <= 0:
+        raise ValueError
```
This line may have been moved from elsewhere but probably should be something like:
  ValueError(f'n ({n}) must be positive')

mdhaber replied:
Thanks. Fixed!

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -549,6 +549,42 @@ def _spearmanr_2cols(x):
+    if n <= 0:
+        raise ValueError
+    elif c < 0 or 4*c > n*(n-1):
+        raise ValueError
```
This line probably should be something like:
  ValueError(f'c ({c}) must satisfy 0 <= 4*c <= n*(n-1) = {n*(n-1)}.')

mdhaber replied:
Thanks. Fixed!

mdhaber commented on scipy/stats/mstats_basic.py:
```diff
@@ -549,6 +549,42 @@ def _spearmanr_2cols(x):
+def _kendall_p_exact(n, c):
+    # Exact p-value, see Maurice G. Kendall, "Rank Correlation Methods" (4th Edition), Charles Griffin & Co., 1970.
+    if n <= 0:
+        raise ValueError
```
```suggestion
        raise ValueError('n ({n}) must be positive')
```
If the user could get here, I would want to provide how `n` relates to their input, but empty `x` and `y` is already checked, so this is fine.

mdhaber commented on scipy/stats/mstats_basic.py:
```diff
@@ -549,6 +549,42 @@ def _spearmanr_2cols(x):
+    if n <= 0:
+        raise ValueError
+    elif c < 0 or 4*c > n*(n-1):
+        raise ValueError
```
```suggestion
        raise ValueError(f'c ({c}) must satisfy 0 <= 4c <= n(n-1) = {n*(n-1)}.')
```

pvanmulbregt commented:
The failure in scipy/linalg/tests/test_decomp_cossin.py::test_cossin_separate appears unrelated to this change.

mdhaber commented:
@pvanmulbregt if you'd like me to take a look, would you be able to send me the relevant parts of Rank Correlation Methods so I can see where these changes come from? If you're comfortable merging though, that's fine, too!

pvanmulbregt commented:
@mdhaber I also don't have access to Rank Correlation Methods.  The change in this PR is to replace a final division by n!/2 with dividing by 3, 4, ...,n, one division each iteration, in order to prevent overflow.  I.e. It is a numerical computation change, not a statistics change.

pvanmulbregt commented:
Thanks @Konrad0 and @mdhaber!

mdhaber commented:
Ah, ok. Thanks!

Konrad0 commented:
Sorry for being absent for such a long time.
Thanks @mdhaber for finishing this PR, much appreciated!
Thanks @pvanmulbregt for the thorough review!
